collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1446,38 +1444,42 @@ private void showNextView() {
                 return;
             }
 
-            formController.getAuditEventLogger().flush();    // Close events waiting for an end time
-
-            switch (event) {
-                case FormEntryController.EVENT_QUESTION:
-                case FormEntryController.EVENT_GROUP:
-                    // create a savepoint
-                    nonblockingCreateSavePointData();
-                    next = createView(event, true);
-                    showView(next, AnimationType.RIGHT);
-                    break;
-                case FormEntryController.EVENT_END_OF_FORM:
-                case FormEntryController.EVENT_REPEAT:
-                case EVENT_PROMPT_NEW_REPEAT:
-                    next = createView(event, true);
-                    showView(next, AnimationType.RIGHT);
-                    break;
-                case FormEntryController.EVENT_REPEAT_JUNCTURE:
-                    Timber.i("Repeat juncture: %s", formController.getFormIndex().getReference());
-                    // skip repeat junctures until we implement them
-                    break;
-                default:
-                    Timber.w("JavaRosa added a new EVENT type and didn't tell us... shame on them.");
-                    break;
-            }
+            animateToNextView();
         } catch (JavaRosaException e) {
             Timber.d(e);
             createErrorDialog(e.getCause().getMessage(), DO_NOT_EXIT);
         }
     }
 
-    private boolean saveBeforeMovingForward(FormController formController) {
-        if (formController != null && formController.currentPromptIsQuestion()) {
+    private void animateToNextView() {
+        int event = getFormController().getEvent();
+
+        switch (event) {
+            case FormEntryController.EVENT_QUESTION:
+            case FormEntryController.EVENT_GROUP:
+                // create a savepoint
+                nonblockingCreateSavePointData();
+                showView(createView(event, true), AnimationType.RIGHT);
+                break;
+            case FormEntryController.EVENT_END_OF_FORM:
+            case FormEntryController.EVENT_REPEAT:
+            case EVENT_PROMPT_NEW_REPEAT:
+                showView(createView(event, true), AnimationType.RIGHT);
+                break;
+            case FormEntryController.EVENT_REPEAT_JUNCTURE:
+                Timber.i("Repeat juncture: %s", getFormController().getFormIndex().getReference());
+                // skip repeat junctures until we implement them
+                break;
+            default:
+                Timber.w("JavaRosa added a new EVENT type and didn't tell us... shame on them.");
+                break;
+        }
+    }
+
+    private boolean saveBeforeNextView(FormController formController) {
+        formController.getAuditEventLogger().flush();    // Close events waiting for an end time
+
+        if (formController.currentPromptIsQuestion()) {
```

Comment:
Yeah I think I need to have a look through these. I really don't like that so many of these methods have to deal with the nullability though. What I might do is try and push that decision as early as I can and use @NonNull and @Nullable to make sure it's clear when we have to check.

Reply:
Yes, I agree with that general approach. Perhaps it doesn't need to be done here but then I might suggest keeping the existing null checks. It would be really ideal if after the form is expected to be loaded there was no way for the form controller to be null. Even better, as we've previously discussed, would be if the form controller could never be null and instead its functionality changed whether or not a form was active but that's further out.

Reply:
So for the moment I'm just going to go with the `FormController` being `@Nullable` in all the `ViewModel` classes that depend on it and have null check guards on all the actions. There is definitely something better on the tip of my tongue - maybe the ViewModels shouldn't be created until after the `FormController` is available or maybe all the ViewModels should talk to some loader/repo that gives access to the `FormController` as a `Future`. I think the right™️ solution is going to need us to move "form loading" away from the `FormEntryActivity` so best to leave it right now.

The `@Nullable` annotation gives us some safety here but ideally we'd write tests for ALL the null cases. I'm very tempted to jump to Kotlin here if I'm being frank as we'd get compile checks and also much nicer sugar for dealing with the nulls here. That would definitely be a different PR but I think it's worth bringing up to get the discussion going.

Reply:
All sound good.